File: Leetcode/Weekly-Contest-378/Find Longest Special Substring That Occurs Thrice II.cpp
class Solution { 
public: 
    int maximumLength(string s) { 
        int ans=0; 
        vector<vector<int>>freq(26); 
        for(int i=0;i<s.size();i++) 
        { 
            int j=i; 
            while(j<s.size() and s[j]==s[i])j++; 
            freq[s[i]-'a'].push_back(j-i); 
            i=j-1; 
        } 
        for(auto x:freq) 
        { 
            if(x.empty())continue; 
            sort(x.begin(),x.end()); 
            int sz=x.size(); 
            ans=max(ans,x.back()-2); 
            if(sz>1) 
            { 
                ans=max(ans,min(x[sz-1]-1,x[sz-2])); 
            } 
            if(sz>2) 
            { 
                ans=max(ans,x[sz-3]); 
            } 
        } 
        if(ans==0)ans=-1; 
        return ans; 
    } 
};

File: Leetcode/Weekly-Contest-378/Find Longest Special Substring That Occurs Thrice II.java
class Solution {
    public int maximumLength(String s) {
        int n = s.length();
        int[][] freq = new int[26][n+1];
        char pre = s.charAt(0);
        
        int len = 1;
        freq[s.charAt(0)-'a'][1] = 1;
        int ans = -1;
        for (int i = 1; i<n; i++) {
            char cur = s.charAt(i);
            if (cur == pre) {
                len++;
                int count = 1;
                freq[cur-'a'][len] += 1;


            } else {
                freq[cur-'a'][1] += 1;
   
                pre = cur;
                len = 1;
            }
        }
        for (int i = 0; i<26; i++) {
            int preSum = 0;
            for (int j = n; j>=1; j--) {
                preSum += freq[i][j];
                if (preSum >= 3) {
                    ans=  Math.max(ans, j);
                    break;
                }
            }
        }
   
        return ans;
    }
}
